micro_http/src/server.rs

```diff
@@ -0,0 +1,455 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+extern crate epoll;
+
+use std::collections::VecDeque;
+use std::io::Write;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::os::unix::net::{UnixListener, UnixStream};
+
+use common::Version;
+pub use common::{ConnectionError, RequestError};
+use connection::HttpConnection;
+use request::Request;
+use response::{Response, StatusCode};
+use std::collections::HashMap;
+
+static SERVER_FULL_ERROR: &'static [u8] = b"HTTP/1.1 500\r\n \
+                                            Server: Firecracker API\r\n \
+                                            Content-Length: 14\r\n\r\nServer is full";
+
+/// Errors pertaining to the server.
+#[derive(Debug)]
+pub enum ServerError {
+    /// Epoll operations failed.
+    IOError,
+    /// Error from one of the connections.
+    ConnectionError(ConnectionError),
+}
+
+enum StreamDirection {
+    Incoming,
+    Outgoing,
+}
+
+/// HTTP Server implementation using Unix Domain Sockets.
+pub struct HttpServer {
+    socket: UnixListener,
+    epoll_fd: RawFd,
+    max_connections: usize,
+    epoll_structs: HashMap<i32, (RawFd, HttpConnection<UnixStream>)>,
+    token_queue: VecDeque<i32>,
+}
+
+impl HttpServer {
+    /// Constructor for `HttpServer`.
+    pub fn new(path_to_socket: String) -> Result<Self, ServerError> {
+        let socket = UnixListener::bind(path_to_socket).map_err(|_| ServerError::IOError)?;
+        let epoll_fd = epoll::create(false).map_err(|_| ServerError::IOError)?;
+        Ok(HttpServer {
+            socket,
+            epoll_fd,
+            max_connections: 10,
+            epoll_structs: HashMap::new(),
+            token_queue: (0..10).collect(),
+        })
+    }
+
+    /// Starts the HTTP Server.
+    pub fn start_server(&mut self) -> Result<(), ServerError> {
+        epoll::ctl(
+            self.epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.socket.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, 100),
+        )
+        .map_err(|_| ServerError::IOError)?;
+
+        Ok(())
+    }
+
+    /// Calls `epoll::wait` on the epoll structure containing all open connections.
+    pub fn wait_server(&mut self) -> Result<Vec<(i32, Request)>, ServerError> {
```

Comment:
I suggest declaring a type synonym:
```
type Token = i32
```
So that it's clearer what this `i32` represents, and also perhaps a tuple struct:
```
struct ParsedRequest(Token, Request);
```
So that the meaning of the return type is clearer.